common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+            this._cbs.attach(callback);
+          },
+          putToken: (audience, token, callback) => {
+            this._cbs.putToken(audience, token, callback);
+          },
+          send: (message: Message, endpoint: string, to: string, done: GenericAmqpBaseCallback<any>): void => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_006: [The `send` method shall construct an AMQP message using information supplied by the caller, as follows:
+            The ‘to’ field of the message should be set to the ‘to’ argument.
+            The ‘body’ of the message should be built using the message argument.] */
+            debug('call to deprecated api \'azure-iot-amqp-base.Amqp.send\'. You should be using SenderLink.send instead');
+            let amqpMessage = AmqpMessage.fromMessage(message);
+            if (to !== undefined) {
+              amqpMessage.properties.to = to;
+            }
+
+            if (!this._senders[endpoint]) {
+              this._fsm.handle('attachSenderLink', endpoint, null, (err) => {
+                if (err) {
+                  debug('failed to attach the sender link: ' + err.toString());
+                  done(err);
+                } else {
+                  (this._senders[endpoint] as SenderLink).send(amqpMessage, done);
+                }
+              });
+            } else {
+              (this._senders[endpoint] as SenderLink).send(amqpMessage, done);
+            }
+
+          },
+          getReceiver: (endpoint: string, done: GenericAmqpBaseCallback<ReceiverLink>): void => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_010: [If a receiver for this endpoint doesn’t exist, the getReceiver method should create a new AmqpReceiver object and then call the done() method with the object that was just created as an argument.] */
+            if (!this._receivers[endpoint]) {
+              this._fsm.handle('attachReceiverLink', endpoint, null, (err) => {
+                if (err) {
+                  done(err);
+                } else {
+                  done(undefined, this._receivers[endpoint]);
+                }
+              });
+            } else {
+              /*Codes_SRS_NODE_COMMON_AMQP_16_009: [If a receiver for this endpoint has already been created, the getReceiver method should call the done() method with the existing instance as an argument.] */
+              this._safeCallback(done, null, this._receivers[endpoint]);
+            }
+          },
+          attachReceiverLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback<ReceiverLink>): void => {
+            debug('creating receiver link for: ' + endpoint);
+            this._receivers[endpoint] = new ReceiverLink(endpoint, linkOptions, this._amqp);
+            const attachErrorHandler = (err) => {
+              debug('receiver link error while attaching: ' + endpoint + ': ' + err.toString());
+              delete(this._receivers[endpoint]);
+              done(err);
```

Comment:
>done [](start = 14, length = 4)

_safeCallback?

Reply:
callbacks are not optional here

---
In reply to: [139525728](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139525728) [](ancestors = 139525728)